_posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd

```diff
@@ -0,0 +1,347 @@
+---
+title: "Introducing torch autograd"
+description: > 
+ Instructive though it may be to code backpropagation from scratch, with torch, there is hardly ever a reason to. Its automatic differentiation feature, called autograd, keeps track of operations that need their gradients computed, as well as how to compute them.
+ In this second post of a four-part series, we update our simple, hand-coded network to make use of autograd.
+author:
+  - name: Sigrid Keydana
+    affiliation: RStudio
+    affiliation_url: https://www.rstudio.com/
+slug: keydanatorchnetworkwithautograd
+date: 10-05-2020
+categories:
+  - Torch
+  - R
+output:
+  distill::distill_article:
+    self_contained: false
+    toc: true
+preview: images/preview.jpg
+editor_options: 
+  markdown: 
+    wrap: 72
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+Last week, we saw how to code [a simple network from
+scratch](https://blogs.rstudio.com/ai/posts/2020-10-01-torch-network-from-scratch),
+using nothing but `torch` *tensors*. Predictions, loss, gradients,
+weight updates -- all these things we've been computing ourselves.
+Today, we make a significant change: Namely, we spare ourselves the
+cumbersome calculation of gradients, and have `torch` do it for us.
+
+Prior to that though, let's get some background.
+
+## Automatic differentiation with *autograd*
+
+`Torch` uses a module called *autograd* to
+
+(1) record operations performed on tensors, and
+
+(2) store what will have to be done to obtain the corresponding
+    gradients, once we're entering the backward pass.
+
+These prospective actions are stored internally as functions, and when
+it's time to compute the gradients, these functions are applied in
+order: Application starts from the output node, and calculated gradients
+are successively being *propagated* *back* through the network. This is
+a form of *reverse mode automatic differentiation*.
+
+#### *Autograd* basics
+
+As users, we can see a bit of the implementation. As a prerequisite for
+this "recording" to happen, tensors have to be created with
+`requires_grad = TRUE`. E.g.,
+
+```{r}
+library(torch)
+
+x <- torch_ones(2, 2, requires_grad = TRUE)
+```
+
+To be clear, `x` now is a tensor *with respect to which* gradients have
+to be calculated -- normally, a tensor representing a weight or a bias,
+not the input data [^1]. If subsequently we perform some operation on
```

Comment:
```suggestion
not the input data [^1]. If we subsequently perform some operation on
```

Reply:
As is, it should be offset with commas, so I just inlined it so that's unnecessary.